guidance/rust_code/src/graph/mod.rs

```diff
@@ -1,402 +1,282 @@
     }
 }
 
-impl PartialEq for Vertex {
-    fn eq(&self, other: &Vertex) -> bool {
-        if self.name == other.name && self.gps_point == other.gps_point {
-            return true;
-        }
+impl TangentialPoint {
+    pub fn distance(&self, other: &Self) -> f64 {
+        let (x, y, z) = other - self;
 
-        false
+        (x.powi(2) + y.powi(2) + z.powi(2)).sqrt()
     }
 }
 
-/// Returns (number_of_edges, number_of_vertices) in the Buffer of KML contents
-pub(self) fn number_of_edges_and_vertices_from_buffer(reader: &mut BufReader<File>) -> (u32, u32) {
-    reader.seek(SeekFrom::Start(0)).unwrap();
+// Clone trait only for unit testing
+#[derive(Debug, Clone)]
+pub struct GPSPointDeg {
+    pub lat: f64,
+    pub long: f64,
+    pub height: f64
+}
 
-    let mut number_of_edges = 0;
-    let mut number_of_vertices = 0;
+impl PartialEq for GPSPointDeg {
+    fn eq(&self, other: &Self) -> bool {
+        self.lat == other.lat 
+            &&
+        self.long == other.long
+    }
+}
 
-    let mut graph_element_found = false;
-    for line in reader.lines() {
-        let line = line.unwrap();
+pub(self) struct GPSPointRad {
+    pub(self) lat: f64,
+    pub(self) long: f64,
+    pub(self) height: f64
+}
 
-        if line.contains("<Placemark>") {
-            graph_element_found = true;
-        } else if line.contains("</Placemark>") {
-            graph_element_found = false;
-        }
+#[derive(Debug, PartialEq)]
+pub struct Edge {
+    // For debugging, finding out what line we are looking at on the map
+    pub name: String,
 
-        if graph_element_found {
-            // Then determine whether its edge or vertex
-            if line.contains("<name>Line") {
-                number_of_edges += 1;
-            } else if line.contains("<name>Point") || line.contains("<name>") {
-                // Covers case of a regular vertex, or a key point in our graph
-                number_of_vertices += 1;
-            }
-        }
-    }
+    // T, in this case, can be a point relative to our tangential frame, or the gps frame
+    pub points: Vec<Point>,
 
-    return (number_of_edges, number_of_vertices);
+    pub distance: f64,
 }
 
-pub(self) fn add_gps_points_to_edges(
-    reader: &mut BufReader<File>,
-    edges: &mut Vec<Edge>,
-) {
-    // Rewind and start from beginning of contents of buffer
-    reader.seek(SeekFrom::Start(0)).unwrap();
+impl Edge {
+    fn new(name: String, points: Vec<Point>) -> Self {
 
-    // buffer for reading lines
-    let mut line = String::new();
+        // Determining the distance of the edge for construction
+        let mut distance = 0.0;
 
-    let mut edge_found = false;
+        let poitns_n = points.iter().map(|point| &point.tangential);
+        let mut points_n_plus_1 = points.iter().map(|point| &point.tangential);
+        points_n_plus_1.next();
 
-    // Items needed for Edge
-    let mut name = String::new();
+        let points_n_n_plus_1 = poitns_n.zip(points_n_plus_1);
```

Comment:
Generally I would appreciate more comments to make it easier for me (who pretends to know rust) to read and review the code without having to manually parse through unfamiliar syntax to figure out what you're trying to do. These iterator operations could be explained in a comment, especially since all the variable names are close to identical as well

Reply:
fixed.